ipa-core/src/protocol/ipa_prf/prf_sharding/mod.rs

```diff
@@ -306,27 +341,42 @@ pub trait GroupingKey {
     fn get_grouping_key(&self) -> u64;
 }
 
-#[tracing::instrument(name = "compute_histogram_of_users_with_row_count", skip_all)]
-pub fn compute_histogram_of_users_with_row_count<S>(input: &[S]) -> Vec<usize>
+#[tracing::instrument(name = "compute_histogram_range_sort_key_per_user", skip_all)]
+pub fn compute_histogram_range_sort_key_per_user<BK, TV, TS>(
+    input: &mut [PrfShardedIpaInputRow<BK, TV, TS>],
+) -> (Vec<usize>, Vec<Range<usize>>)
 where
-    S: GroupingKey,
+    BK: SharedValue,
+    TV: SharedValue,
+    TS: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
 {
     let mut histogram = vec![];
     let mut last_prf = input[0].get_grouping_key() + 1;
     let mut cur_count = 0;
-    for row in input {
+    let mut start = 0;
+    let mut ranges = vec![];
```

Comment:
Can we allocate a vector with capacity? Do we have any way to estimate the total number of users?

Reply:
Does this have to be done all at once?  It seems like it could be sharded, even within a shard.  (Starting with a single vector is probably fine for now, of course.)

Reply:
The sort code which @benjaminsavage rewrote assumes that ALL the rows will come together to be sorted. Unfortunately, this takes away the opportunity to shard the sorting process

Reply:
No, it's totally possible to shard this. You could make multiple calls to quicksort- and just provide each one with a different set of ranges. This could either be done across multiple machines, or multiple sequential calls from the same machine. We can probably convert this to a stream of ranges even since I just get an iterator and map it.

Reply:
wont it also need having a different step name (if it happens on the same machine) since it might conflict with what infra has already seen?

Reply:
> Can we allocate a vector with capacity? Do we have any way to estimate the total number of users?

Worst case it could = number of input rows (if all rows are for unique users). No way of doing it until this part of the code.

Reply:
Yes, you would need to have a step that sits above the sharding.  Or even completely new contexts (which I'd avoid).  But you aren't creating an unbounded number of shards, so that works within our current framework.